text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+meaningfully different behavior is considered unlikely enough that the RFC is
+willing to permit it to be labeled as a minor change -- and otherwise, inherent
+methods could never be added after the fact.
+
+### Other items
+
+Most remaining items do not have any particularly unique items:
+
+* For type aliases, see "[Signatures in type definitions](#signatures-in-type-definitions)".
+* For free functions, see "[Signatures in functions](#signatures-in-functions)".
+
+## Cross-cutting concerns
+
+### Behavioral changes
+
+This RFC is largely focused on API changes which may, in particular, cause
+downstream code to stop compiling. But in some sense it is even more pernicious
+to make a change that allows downstream code to continue compiling, but causes
+its runtime behavior to break.
+
+This RFC does not attempt to provide a comprehensive policy on behavioral
+changes, which would be extremely difficult. In general, APIs are expected to
+provide explicit contracts for their behavior via documentation, and behavior
+that is not part of this contract is permitted to change in minor
+revisions. (Remember: this RFC is about setting a *minimum* bar for when major
+version bumps are required.)
+
+This policy will likely require some revision over time, to become more explicit
+and perhaps lay out some best practices.
+
+### Signatures in type definitions
+
+#### Major change: tightening bounds.
+
+Adding new constraints on existing type parameters is a breaking change, since
+existing uses of the type definition can break. So the following is a major
+change:
+
+```rust
+// MAJOR CHANGE
+
+// Before
+struct Foo<A> { .. }
+
+// After
+struct Foo<A: Clone> { .. }
+```
+
+#### Minor change: loosening bounds.
+
+Loosening bounds, on the other hand, cannot break code because when you
+reference `Foo<A>`, you *do not learn anything about the bounds on `A`*. (This
+is why you have to repeat any relevant bounds in `impl` blocks for `Foo`, for
+example.) So the following is a minor change:
+
+```rust
+// MINOR CHANGE
+
+// Before
+struct Foo<A: Clone> { .. }
+
+// After
+struct Foo<A> { .. }
+```
+
+#### Minor change: adding defaulted type parameters.
+
+All existing references to a type/trait definition continue to compile and work
+correctly after a new defaulted type parameter is added. So the following is
+a minor change:
+
+```rust
+// MINOR CHANGE
+
+// Before
+struct Foo { .. }
+
+// After
+struct Foo<A = u8> { .. }
+```
+
+#### Minor change: generalizing to generics.
+
+A struct or enum field can change from a concrete type to a generic type
+parameter, provided that the change results in an identical type for all
+existing use cases. For example, the following change is permitted:
+
+```rust
+// MINOR CHANGE
+
+// Before
+struct Foo(pub u8);
+
+// After
+struct Foo<T = u8>(pub T);
+```
+
+because existing uses of `Foo` are shorthand for `Foo<u8>` which yields the
+identical field type.
```

Comment:
This is not actually the case right now since #213 isn't fully implemented:
```rust
enum Foo<S = i32> {
    Bar,
    Baz(S)
}

fn f<S>(_: Foo<S>) {}

fn main() {
    f(Foo::Bar);
}
```

```
~ ❯ rustc test.rs
test.rs:9:5: 9:6 error: unable to infer enough type information about `_`; type annotations required [E0282]
test.rs:9     f(Foo::Bar);
              ^
error: aborting due to previous error
```